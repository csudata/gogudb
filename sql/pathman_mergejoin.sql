\set VERBOSITY terse

SET search_path = 'public','_gogu';
CREATE EXTENSION gogudb ;
CREATE SCHEMA test;


CREATE TABLE test.range_rel (
	id	SERIAL PRIMARY KEY,
	dt	TIMESTAMP NOT NULL,
	txt	TEXT);
CREATE INDEX ON test.range_rel (dt);

INSERT INTO test.range_rel (dt, txt)
SELECT g, md5(g::TEXT) FROM generate_series('2015-01-01', '2015-04-30', '1 day'::interval) as g;

SELECT _gogu.create_range_partitions('test.range_rel', 'DT', '2015-01-01'::DATE, '1 month'::INTERVAL);

CREATE TABLE test.num_range_rel (
	id	SERIAL PRIMARY KEY,
	txt	TEXT);

INSERT INTO test.num_range_rel SELECT g, md5(g::TEXT) FROM generate_series(1, 3000) as g;

SELECT _gogu.create_range_partitions('test.num_range_rel', 'id', 0, 1000, 4);

/*
 * Merge join between 3 partitioned tables
 *
 * test case for the fix of sorting, merge append and index scan issues
 * details in commit 54dd0486fc55b2d25cf7d095f83dee6ff4adee06
 */
SET enable_hashjoin = OFF;
SET enable_nestloop = OFF;
SET enable_mergejoin = ON;

EXPLAIN (COSTS OFF)
SELECT * FROM test.range_rel j1
JOIN test.range_rel j2 on j2.id = j1.id
JOIN test.num_range_rel j3 on j3.id = j1.id
WHERE j1.dt < '2015-03-01' AND j2.dt >= '2015-02-01' ORDER BY j2.dt;

SET enable_hashjoin = ON;
SET enable_nestloop = ON;


DROP SCHEMA test CASCADE;
DROP EXTENSION gogudb;

