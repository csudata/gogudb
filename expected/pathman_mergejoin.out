\set VERBOSITY terse
SET search_path = 'public','_gogu';
CREATE EXTENSION gogudb ;
CREATE SCHEMA test;
CREATE TABLE test.range_rel (
	id	SERIAL PRIMARY KEY,
	dt	TIMESTAMP NOT NULL,
	txt	TEXT);
CREATE INDEX ON test.range_rel (dt);
INSERT INTO test.range_rel (dt, txt)
SELECT g, md5(g::TEXT) FROM generate_series('2015-01-01', '2015-04-30', '1 day'::interval) as g;
SELECT _gogu.create_range_partitions('test.range_rel', 'DT', '2015-01-01'::DATE, '1 month'::INTERVAL);
 create_range_partitions 
-------------------------
                       4
(1 row)

CREATE TABLE test.num_range_rel (
	id	SERIAL PRIMARY KEY,
	txt	TEXT);
INSERT INTO test.num_range_rel SELECT g, md5(g::TEXT) FROM generate_series(1, 3000) as g;
SELECT _gogu.create_range_partitions('test.num_range_rel', 'id', 0, 1000, 4);
 create_range_partitions 
-------------------------
                       4
(1 row)

/*
 * Merge join between 3 partitioned tables
 *
 * test case for the fix of sorting, merge append and index scan issues
 * details in commit 54dd0486fc55b2d25cf7d095f83dee6ff4adee06
 */
SET enable_hashjoin = OFF;
SET enable_nestloop = OFF;
SET enable_mergejoin = ON;
EXPLAIN (COSTS OFF)
SELECT * FROM test.range_rel j1
JOIN test.range_rel j2 on j2.id = j1.id
JOIN test.num_range_rel j3 on j3.id = j1.id
WHERE j1.dt < '2015-03-01' AND j2.dt >= '2015-02-01' ORDER BY j2.dt;
                                         QUERY PLAN                                          
---------------------------------------------------------------------------------------------
 Sort
   Sort Key: j2.dt
   ->  Merge Join
         Merge Cond: (j2.id = j3.id)
         ->  Merge Join
               Merge Cond: (j1.id = j2.id)
               ->  Merge Append
                     Sort Key: j1.id
                     ->  Index Scan using _test_1_range_rel_pkey on _test_1_range_rel j1
                     ->  Index Scan using _test_2_range_rel_pkey on _test_2_range_rel j1_1
               ->  Merge Append
                     Sort Key: j2.id
                     ->  Index Scan using _test_2_range_rel_pkey on _test_2_range_rel j2
                     ->  Index Scan using _test_3_range_rel_pkey on _test_3_range_rel j2_1
                     ->  Index Scan using _test_4_range_rel_pkey on _test_4_range_rel j2_2
         ->  Append
               ->  Index Scan using _test_1_num_range_rel_pkey on _test_1_num_range_rel j3
               ->  Index Scan using _test_2_num_range_rel_pkey on _test_2_num_range_rel j3_1
               ->  Index Scan using _test_3_num_range_rel_pkey on _test_3_num_range_rel j3_2
               ->  Index Scan using _test_4_num_range_rel_pkey on _test_4_num_range_rel j3_3
(20 rows)

SET enable_hashjoin = ON;
SET enable_nestloop = ON;
DROP SCHEMA test CASCADE;
NOTICE:  drop cascades to 12 other objects
DROP EXTENSION gogudb;
